Streamline dbt test creation [DE-222]

jackwelty opened the pull request:
This PR adds a handful of features to our dbt schema.yml generation script.

1. Merge / overwrite capabilities - the script can now be run with `--merge` and `--overwrite` flags, so that new tables / columns / tests can be added to existing `schema.yml` files, without overwriting existing work. Previously this script could only overwrite files, meaning any additional custom test configuration would be wiped out. Going forward, we should re-run our schema gen script any time we add new models or columns to an existing sync.
2. "Universal test" capability - Users can now pass in a yml file of "universal tests", i.e. tests that should be configured on any column in the project matching a specific name. For example, any time we see `_airbyte_ab_id`, we should test to ensure it is unique and not null. There is a default file at `utils/universal_tests.yml`, but the script takes a filename as an argument, so users can specify a new file per sync.
3. Incorporate into `init_dbt.sh` script - this python script now runs at the end of `init_dbt.sh`, so that only one command is needed to take Airbyte's autogenerated dbt and transform it into a standalone dbt project, with inital tests!

emily-flambe commented:
This is VERY exciting - Don't see any issues with the code, and it just so happens that I have a ticket this sprint that gives me an opportunity to take this for a proper test drive 👀 which I will do before rendering a proper Review here, to see if anything comes up in the process 👀 👀

jackwelty commented on utils/README.md:
````diff
@@ -74,25 +75,39 @@ You can use the same command - just replace the ```$(gcloud compute instance-gro
+yaml file. It queries the database to get the column names, and then uses the
+`universal_tests.yml` file to generate a set of tests for each column.
+
+This script runs at the end of `init_dbt`, but can also be run independently.
````
added!

emily-flambe replied:
ahhhh exquisite!!

emily-flambe commented:
hmmm @jackwelty I'm running into a couple of bugs when trying to run the script on the dbt I just initialized for Empower (which I pushed to branch [ec/jw/add-schemayml-merge](https://github.com/community-tech-alliance/dbt-cta/tree/ec/jw/add-schemayml-merge)).

First, it doesn't seem to appreciate the --sync-name argument:

```emilycogsdill@MacBook-Pro-3 dbt-cta % python3.9 ../utils/generate_schema_yml.py --sync-name empower
usage: generate_schema_yml.py [-h] [--sync_name SYNC_NAME] [--merge] [--overwrite]
                              [--universal-tests UNIVERSAL_TESTS]
generate_schema_yml.py: error: unrecognized arguments: --sync-name empower
```

And when I just set my env variable `SYNC_NAME=empower` without using that argument, I get a different error instead:

```emilycogsdill@MacBook-Pro-3 dbt-cta % python3.9 ../utils/generate_schema_yml.py
Traceback (most recent call last):
  File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 256, in <module>
    main(args)
  File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 212, in main
    dirs = list_model_directories(args.sync_name)
  File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 36, in list_model_directories
    d for d in [base + "/" + b for b in os.listdir(base)] if os.path.isdir(d)
FileNotFoundError: [Errno 2] No such file or directory: 'None/models'
```

Can you try checking out my branch and running for yourself to see if you can see what's up? 🙏

emily-flambe commented:
> Ok one of these is a dumb hyphen/underscore typo from yours truly ([pushed a fix to your branch](https://github.com/community-tech-alliance/dbt-cta/commit/ea77e8d153f1af63d035e39cb1feb0e51151143a)).

Ahh! yay, that bit is indeed fixed 🙌 

> Beyond that, I'm still hitting an error because within empower/models/1_cta_full_refresh, my script is expecting to find a bunch of SQL files, but instead everything is nested one additional level in empower_partner_a. Do we expect that models will exist inside this additional subdirectory? I can add some looping to handle this, but all of our other syncs just have [sync_name]/models/[model_type]/[actual sql files].

Ooops that was my bad! I just moved those out of the nested folder (indeed that should not be the case, twas an oversight on my part). Now I am at least hitting a different error:

```
emilycogsdill@MacBook-Pro-3 dbt-cta % python3.9 ../utils/generate_schema_yml.py --sync-name empower
Traceback (most recent call last):
  File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 256, in <module>
    main(args)
  File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 215, in main
    schema_dict = generate_schema_dict(d)
  File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 129, in generate_schema_dict
    model_configs = [get_model_config(m, target) for m in models]
  File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 129, in <listcomp>
    model_configs = [get_model_config(m, target) for m in models]
  File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 61, in get_model_config
    model_config = output.split("models:")[1]
```

🤔

jackwelty commented:
> > Ok one of these is a dumb hyphen/underscore typo from yours truly ([pushed a fix to your branch](https://github.com/community-tech-alliance/dbt-cta/commit/ea77e8d153f1af63d035e39cb1feb0e51151143a)).
> 
> Ahh! yay, that bit is indeed fixed 🙌
> 
> > Beyond that, I'm still hitting an error because within empower/models/1_cta_full_refresh, my script is expecting to find a bunch of SQL files, but instead everything is nested one additional level in empower_partner_a. Do we expect that models will exist inside this additional subdirectory? I can add some looping to handle this, but all of our other syncs just have [sync_name]/models/[model_type]/[actual sql files].
> 
> Ooops that was my bad! I just moved those out of the nested folder (indeed that should not be the case, twas an oversight on my part). Now I am at least hitting a different error:
> 
> ```
> emilycogsdill@MacBook-Pro-3 dbt-cta % python3.9 ../utils/generate_schema_yml.py --sync-name empower
> Traceback (most recent call last):
>   File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 256, in <module>
>     main(args)
>   File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 215, in main
>     schema_dict = generate_schema_dict(d)
>   File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 129, in generate_schema_dict
>     model_configs = [get_model_config(m, target) for m in models]
>   File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 129, in <listcomp>
>     model_configs = [get_model_config(m, target) for m in models]
>   File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 61, in get_model_config
>     model_config = output.split("models:")[1]
> ```
> 
> 🤔

I just cloned your branch again and was able to run schema-gen? I pushed my [generated yml](https://github.com/community-tech-alliance/dbt-cta/commit/17010d8e9150eeb3e3efb7db16919bc7393bb3a4) but feel free to delete.
<img width="592" alt="image" src="https://user-images.githubusercontent.com/51384930/231188401-6800f360-676f-4b81-a33c-2e0dd8ec17ce.png">

emily-flambe commented:
This is so mysterious! I temporarily removed the yml you pushed and tried running again, still hitting this error: 

```

emilycogsdill@MacBook-Pro-3 dbt-cta % pipenv run python ../utils/generate_schema_yml.py --sync-name empower
Traceback (most recent call last):
  File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 256, in <module>
    main(args)
  File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 215, in main
    schema_dict = generate_schema_dict(d)
  File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 129, in generate_schema_dict
    model_configs = [get_model_config(m, target) for m in models]
  File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 129, in <listcomp>
    model_configs = [get_model_config(m, target) for m in models]
  File "/Users/emilycogsdill/Documents/GitHub/dbt-cta/dbt-cta/../utils/generate_schema_yml.py", line 61, in get_model_config
    model_config = output.split("models:")[1]
IndexError: list index out of range
```

(I have to use `pipenv run python` bc something is scuffed with my global python lol idk)

I am so confused lol. Let's maybe hop on a slack huddle and screen share sometime this afternoon 🤔 🤔 🤔 🤔 🤔 🤔

emily-flambe reviewed (approved):
And what a jolly journey of discovery this was!
